utils/dyn-config.c

```diff
@@ -0,0 +1,597 @@
+    if (!option) {
+      return;
+    }
+  }
+
+  /* parse the int/string type options */
+  switch (option->type) {
+  case GB_OPT_INT:
+    while (!isdigit(*s)) {
+      s++;
+    }
+    r = s;
+    while (isdigit(*r)) {
+      r++;
+    }
+    *r= '\0';
+
+    option->optInt = atoi(s);
+    break;
+  case GB_OPT_STR:
+    while (isblank(*s)) {
+      s++;
+    }
+    /* skip first " or ' if exist */
+    if (*s == '"' || *s == '\'') {
+      s++;
+    }
+    r = q - 1;
+    while (isblank(*r)) {
+      r--;
+    }
+    /* skip last " or ' if exist */
+    if (*r == '"' || *r == '\'') {
+      *r = '\0';
+    }
+
+    /* free if this is reconfig */
+    if (option->optStr) {
+      GB_FREE(option->optStr);
+    }
+    GB_STRDUP(option->optStr, s);
+    break;
+  default:
+    LOG("dyncfg", GB_LOG_ERROR,
+        "option type %d not supported!\n", option->type);
+    break;
+  }
+}
+
+static void
+glusterBlockParseOptions(gbConfig *cfg, char *buf, int len, bool reloading)
+{
+  char *cur = buf, *end = buf + len;
+
+  while (cur < end) {
+    /* skip blanks lines */
+    GB_SKIP_BLANK_LINES(cur, end);
+
+    /* skip comments with '#' */
+    GB_SKIP_COMMENT_LINES(cur, end);
+
+    if (cur >= end) {
+      break;
+    }
+
+    if (!isalpha(*cur)) {
+      continue;
+    }
+
+    /* parse the options from config file to gb_options[] */
+    glusterBlockParseOption(&cur, end);
+  }
+
+  /* parse the options from gb_options[] to struct gbConfig */
+  glusterBlockConfSetOptions(cfg, reloading);
+}
+
+static int
+glusterBlockLoadConfig(gbConfig *cfg, bool reloading)
+{
+  int ret = -1;
+  int fd, len;
+  char *buf;
+
+  buf = CALLOC(GB_MAX_CFG_FILE_SIZE);
+  if (!buf) {
+    return -ENOMEM;
+  }
+
+  fd = open(cfg->configPath, O_RDONLY);
+  if (fd < 0) {
+    LOG("dyncfg", GB_LOG_ERROR,
+        "Failed to open file '%s', %m\n", cfg->configPath);
+    goto free_buf;
+  }
+
+  len = glusterBlockReadConfig(fd, buf, GB_MAX_CFG_FILE_SIZE);
+  close(fd);
+  if (len < 0) {
+    LOG("dyncfg", GB_LOG_ERROR,
+        "Failed to read file '%s'\n", cfg->configPath);
+    goto free_buf;
+  }
+
+  buf[len] = '\0';
+
+  glusterBlockParseOptions(cfg, buf, len, reloading);
+
+  ret = 0;
+free_buf:
+  GB_FREE(buf);
+  return ret;
+}
+
+#define BUF_LEN 1024
+static void *
+glusterBlockDynConfigStart(void *arg)
+{
+  gbConfig *cfg = arg;
+  int monitor, wd, len;
+  char buf[BUF_LEN];
+
+  monitor = inotify_init();
+  if (monitor == -1) {
+    LOG("dyncfg", GB_LOG_ERROR,
+        "Failed to init inotify %d\n", monitor);
+    return NULL;
+  }
+
+  wd = inotify_add_watch(monitor, cfg->configPath, IN_ALL_EVENTS);
+  if (wd == -1) {
+    LOG("dyncfg", GB_LOG_ERROR,
+        "Failed to add \"%s\" to inotify %m\n", cfg->configPath);
+    return NULL;
+  }
+
+  LOG("dyncfg", GB_LOG_INFO,
+      "Inotify is watching \"%s\", wd: %d, mask: IN_ALL_EVENTS\n",
+      cfg->configPath, wd);
+
+  while (1) {
+    struct inotify_event *event;
+    char *p;
```

Comment:
can we move this to beginning of the function ?

Reply:
Yeah, will fix it.

#define GB_MAX_CFG_FILE_SIZE (2 * 1024 * 1024)

Sorry just forgot this one, the size is just calculated the possible size including the commented lines in the /etc/sysconfig/gluster-blockd files.  Or do you have any other ideas about this ? And I am okay with other reasonable size.

Reply:
Read one line by line and move some logic about checking for valid config line (skip blank, commented lines) from glusterBlockParseOptions() to glusterBlockReadConfig(), then sort/fetch the gb_options[] in glusterBlockReadConfig(), then call glusterBlockConfSetOptions() finally :-)

Reply:
Yeah, this make sense.

Do you mind to do a fix in the following PR ?

I am worried the stable of the dyn feature. Since this code is from the tcmu-runner and has been used and test for a very long time, to address this it will change a lot.

If today is not the deadline, I can change this and need to have long time test very possible corner cases.

Reply:
I'm okay, fixing this later, but we are allocating 2 * 1024 * (1024 bytes) = 2 2MiB sole for this. Isn't that too big ? can we atleast shrink this ?

Reply:
Yeah, let's use 32K for now, is that okay ?

Reply:
Yeah, that should be more than sufficient IMO.

Reply:
Sure, done.